docs: the plugin chapter

Exirel opened the pull request:
### Description

I moved the content of `source/plugin.rst` to `source/plugin/*.rst` files and then worked around that structure:

* a glossary
* what is a plugin (explain what it is, plugin types, and plugin name)
* anatomy of a plugin (plugin callables, jobs, setup & shutdown, and about plugin configuration)
* interact with the bot (how to use `bot` to talk, join channels, and stuffs)
* `sopel.plugin` (plugin decorators, already explained in anatomy of a plugin)
* `sopel.plugins` (internals, no need to touch that)

I took the same approach I used for the core configuration: one part is a thorough explanation of the topic and the other part is the good old `autodoc`. I really like it!

I think it's possible to do even more, and that should be done later, either by me or someone else, in this version or a future one. The most important here is that I think it's a good structure to work with.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

Exirel commented:
@dgw if by chance you have the time to review some documentation, I think I'm more or less done with:

* the glossary
* What is a plugin
* Anatomy of a plugin

That last one in particular: I took the same approach I had with the configuration, i.e. "Anatomy of a plugin" is an exploration of everything around a plugin and its rules (which is at the core of any plugin), and I left the autodoc part in another section. I really like how it turned out, but you know how poor my English can be.

The next steps in that documentation PR are probably about how to use the `bot` and the `trigger` arguments, because having to read the whole API documentation for that is not fun, and I've seen many questions about how to do X or Y because the answers are usually buried deep into the code, or fragmented across multiple docstrings.

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
-------------
-.. automodule:: sopel.loader
-   :members:
+=====================
+Plugin for Developers
+=====================
```
```suggestion
===========================
Plugins: Developer Overview
===========================
```

Here's another idea. At least a step in the right direction?

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+
+   Generic rule
+      A generic rule matches any message using a regular expression. It doesn't
+      use any specific format unlikes :term:`Named rule`. It can match the
```
```suggestion
      use any specific format, unlike a :term:`Named rule`. It can match the
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      A generic rule matches any message using a regular expression. It doesn't
+      use any specific format unlikes :term:`Named rule`. It can match the
+      whole message (:term:`Match rule`), or any part of it
+      (:term:`Search rule`), or it may trigger for every matches in the message
```
```suggestion
      (:term:`Search rule`), or it may trigger for every match in the message
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+
+   Named rule
+      A named rule is a rule that uses a regular expression with a specific
+      format: with a name, and usualy with a prefix, or a set of specific
```
```suggestion
      format: with a name, and usually with a prefix, or a set of specific
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      the rule.
+
+   Command
+      A command is a :term:`Named rule` that react to a :term:`Command prefix`
```
```suggestion
      A command is a :term:`Named rule` that reacts to a :term:`Command prefix`
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      and a name.
+
+   Action command
+      An action command is a :term:`Named rule` that react to a name in a
+      message send with the ``ACTION`` intent/CTCP.
```
```suggestion
      An action command is a :term:`Named rule` that reacts to a name in a
      message sent with the ``ACTION`` intent/CTCP.
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      message send with the ``ACTION`` intent/CTCP.
+
+   Nick command
+      A nick command (or nickname command) is a :term:`Named rule` that react
```
```suggestion
      A nick command (or nickname command) is a :term:`Named rule` that reacts
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      :doc:`plugin/anatomy` chapter for the plugin callable signature.
+
+   Plugin job
+      A plugin job is a Python callable, that executes every given period of
+      time. See also the :doc:`plugin/anatomy` chapter for the plugin
+      job signature.
```
```suggestion
      A plugin job is a Python callable, that executes periodically on a
      schedule. See also the :doc:`plugin/anatomy` chapter for the plugin
      job signature.
```

w o r d s m i t h i n g

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+.. glossary::
+   :sorted:
+
+   Sopel Plugin
```
```suggestion
   Sopel plugin
```

Nothing else is Title Case, so why is this?

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      job signature.
+
+   Single file plugin
+      A :term:`Sopel Plugin` composed of a single Python file. Can be loaded by
```
```suggestion
      A :term:`Sopel plugin` composed of a single Python file. Can be loaded by
```

Goes with above suggested change.

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+
+   Folder plugin
+      A plugin composed of a directory that contains a ``__init__.py`` file.
+      It is not considered as a proper Python package unless it's parent
```
```suggestion
      It is not considered as a proper Python package unless its parent
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+   Folder plugin
+      A plugin composed of a directory that contains a ``__init__.py`` file.
+      It is not considered as a proper Python package unless it's parent
+      directory is in ``sys.path``. As it can create confusion, they are not
```
```suggestion
      directory is in ``sys.path``. As they can create confusion, they are not
```

"confusion" might not be the best way to describe this. I hesitate to come right out and say "problems", but that's really the crux of it: folder plugins don't reload well, among other issues.

Exirel replied:
And they are just terrible at absolutely everything anyway. So yeah, let's say "problems".

Exirel replied:
```suggestion
      directory is in ``sys.path``. As they can create problems, they are not
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      It is not considered as a proper Python package unless it's parent
+      directory is in ``sys.path``. As it can create confusion, they are not
+      recommended, and either :term:`Single file plugin` or
+      :term:`Entry point plugin` should be used instead.
```
```suggestion
      :term:`Entry point plugin` structure should be used instead.
```

Like? No?

Exirel replied:
I feel like each "term" is the right label and it doesn't need any addition (like "structure" or "design").

So not a like, but I'm not an English expert, so I can miss some subtlety of the language.

dgw replied:
Fair enough. I was on the fence about this suggestion, and I don't like the added word "better enough" to push.

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+   Namespace package plugin
+      A plugin that is a Python namespace package, i.e. a package within a
+      specific namespace (``sopel_modules.<name>``, where ``sopel_modules`` is
+      the namespace, and ``<name>`` is the Plugin's name). This is the old way
```
```suggestion
      the namespace, and ``<name>`` is the plugin's name). This is the old way
```

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+      A plugin that is an installed Python package and exposed through the
+      ``sopel.plugins`` Python setup entry point named.
+
+   Sopelunking
+      Action performed by a :term:`Sopelunker`.
+
+   Sopelunker
+      A person who does :term:`Sopelunking`.
```
LMAO

dgw commented on docs/source/plugin.rst:
```diff
@@ -1,189 +1,128 @@
+
+   Entry point plugin
+      A plugin that is an installed Python package and exposed through the
+      ``sopel.plugins`` Python setup entry point named.
```
I'm honestly not sure what was meant by this. This suggestion is a stab in the (almost) dark.

```suggestion
      ``sopel.plugins`` setuptools entry point name.
```

Exirel replied:
I've honestly no idea.

```suggestion
      ``sopel.plugins`` setuptools entry point.
```

Exirel replied:
Oh I figured out: my initial sentence might have been "named sopel.plugins", and then I changed that without removing the "named" part.

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+
+Before writing your first plugin, you may want to know what is a plugin, how is
+it made, and what are the best solutions for your need. In this chapter we'll
+cover what is a Plugin, what you can do with them, and some vocabulary.
```
```suggestion
cover what is a plugin, what you can do with them, and some vocabulary.
```

Or cross-reference to the glossary term `Sopel plugin`?

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+it made, and what are the best solutions for your need. In this chapter we'll
+cover what is a Plugin, what you can do with them, and some vocabulary.
+
+You may want to skip ahead to :doc:`anatomy` if you already know what a Plugin
```
```suggestion
You may want to skip ahead to :doc:`anatomy` if you already know what a plugin
```

I'm going on a crusade against these unnecessarily capitalized words, if you couldn't tell. 😹

Exirel replied:
Unlike the ones in Human History, this is a rightful one.

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+-----------
+
+A **Single file** plugin is the most basic form a Sopel plugin can take. It is
+composed of a single file available to Sopel from its plugin's directory (or
```
```suggestion
composed of a single file available to Sopel from its ``plugins`` directory (or
```

?

Exirel replied:
I wanted to say: the directory where Sopel looks for plugins.

dgw replied:
Oh, maybe just `plugins` without the backticks/formatting then.

Exirel replied:
```suggestion
composed of a single file available to Sopel from its plugins directory (or
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+should write a single file plugin when:
+
+* everything is in a single file
+* you don't have any dependencies other than Sopel and the Python's built-in
+  library
```
```suggestion
* you don't have any dependencies other than Sopel and Python's standard library
```

Aside: I don't entirely agree with this—it's perfectly fine to write a single-file plugin with third-party dependencies, in my mind, especially if it's not intended for distribution.

I might add more suggestions about this in the next review round, if you don't vehemently object to explicitly giving users the freedom to skip learning about entry points and packages right away, while they ease into plugin (or Python) development first.

Exirel replied:
Well it's complicated. Probably worth a conversation.

dgw replied:
New alternative suggestion, since we haven't gotten around to having this conversation yet… (sadly)

```suggestion
* you don't have any dependencies other than Sopel and Python's built-in library,
  or your plugin is a "one-off" for your use only (not intended to be shared)
```

I still don't see why we'd mention dependencies for this, but not the `Folder` type.

Exirel replied:
Oh we totally should @ folder plugins!

dgw replied:
Going to do that during rebase before final review?

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+------
+
+A **Folder** plugin is a folder that contains a ``__init__.py`` file. It is
+available to Sopel the same way a :ref:`Single file` plugin is, and share the
+same limitation. Everything that is not available in this ``__init__.py`` file
```
```suggestion
available to Sopel the same way a :ref:`Single file` plugin is, and shares the
same limitation: Everything that is not available in this ``__init__.py`` file
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+so it is not possible to do ``from my_plugin_dir import some_module`` without
+a specific configuration from the bot's owner.
+
+When a Plugin author wants to split code into different files, and/or share it
```
```suggestion
When a plugin author wants to split code into different files, and/or share it
```

🔠🛡️⚔️

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+-----------------
+
+A **Namespace package** plugin is a Python :term:`namespace package`, using
+the namespace ``sopel_modules``. It must installed in the Python's environement
```
```suggestion
the namespace ``sopel_modules``. It must be installed in the Python environment
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+to be used by Sopel, and can require Python dependencies.
+
+Given a ``sopel_modules.plugin`` plugin, Sopel will load everything that is
+available from ``sopel_modules/plugin/__init__.py`` file.
```
```suggestion
available from the ``sopel_modules/plugin/__init__.py`` file.
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+It is the initial version of Sopel's packaged plugins: it can be packaged and
+uploaded to `PyPI`_ then installed using ``pip install``.
+
+When a Plugin author wants to distribute a Sopel plugin, the best option is to
```
```suggestion
When a plugin author wants to distribute a Sopel plugin, the best option is to
```

🔠🛡️⚔️

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+uploaded to `PyPI`_ then installed using ``pip install``.
+
+When a Plugin author wants to distribute a Sopel plugin, the best option is to
+use an :ref:`Entry point` plugin instead.
```
May be worth mentioning older Sopel version compatibility here and the other places entry-point plugins get recommended, since <7 doesn't support entry points.

Exirel replied:
I'll add a note in the `Entry point` section, but I'm not fond of adding it to all the other places. After all, this is the last up-to-date version of the documentation, Sopel 7 has been out for a while now, and this won't be published online until 7.1 comes out. Beside, future version of Sopel going forward will support entry points.

So, agree if it was an archive/a history book, disagree with a documentation that moves forward.

dgw replied:
Ah, true. I guess we can pretty easily get away with basically telling people, "If this doesn't work, your Sopel is too old and you should update it." 😁

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+
+* you want to distribute your plugin on `PyPI`_
+* you want to split the code in multiple files
+* you have dependencies other than Sopel and the Python's built-in library
```
```suggestion
* you have dependencies beyond Sopel and Python's standard library
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+.. __: `Entry points specification`_
+
+
+Plugin's Name
+=============
```
```suggestion
Naming plugins
==============
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,133 @@
+* you want a modern and reliable way to package your Sopel plugin
+* you want to distribute more than one Sopel plugin per distributed package
+
+An entry point plugin is the best way to package and distribute clean and easy
+to update Sopel plugin.
```
```suggestion
An entry point plugin is the best way to package and distribute a Sopel plugin
or collection of plugins in a clean and easy-to-update manner.
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+* :term:`URL callback`: :func:`~sopel.plugin.url` and
+  :func:`~sopel.plugin.url_lazy`
+
+Additionnaly, Sopel identifies a callable as a generic rule when these
+decorator are used alone:
```
```suggestion
Additionally, Sopel identifies a callable as a generic rule when these
decorators are used alone:
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+   def on_join(bot, trigger):
+      pass
+
+Channel vs Private message
+--------------------------
```
```suggestion
Channel vs. private messages
----------------------------
```

I interpret this as `(Channel vs. private) messages`, but am totally open to feedback if it seems more ambiguous than I think.

Exirel replied:
I didn't know how to word that properly, so your suggestion rules.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+better to limit who can trigger them. There are decorators for that:
+
+* :func:`sopel.plugin.require_account`: requires services/NickServ
+  authentication; works only of the server implement authentication properly
```
```suggestion
  authentication; works only if the server implements modern IRC authentication
```

It's not "properly" per se. Just a newer standard that older networks usually don't support.

Exirel replied:
I reused some part of the decorator's docstring for that.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+  authentication; works only of the server implement authentication properly
+* :func:`sopel.plugin.require_privilege`: requires a specific level of
+  privileges in the channel; works only for channel messages, not private
+  messages, and it should be used with :func:`~sopel.plugin.require_chanmsg`
```
This makes me think that we should perhaps consider making `require_privilege` imply `require_chanmsg` without forcing the plugin dev to specify both. Separate PR, of course, but the idea is what's important here.

Exirel replied:
Probably yes.

dgw replied:
Just to follow up, it's already documented as ignored in PMs: https://github.com/sopel-irc/sopel/blob/9189ad01f064987d507938352e1e59bd3cbf128a/sopel/plugin.py#L804

New idea sans wrapping, because I'm short on time:

```suggestion
  messages, and you probably want to use it with :func:`~sopel.plugin.require_chanmsg`
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+often Sopel can send messages to the network. By default, a rule doesn't have
+any rate limiting.
+
+There are 3 type of rate limiting:
```
```suggestion
There are three types of rate limiting:
```

Blah blah style manuals blah blah numbers less than ten.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+   def you_said_ah(bot, trigger):
+      bot.reply('Ha AH!')
+
+Rule label
+----------
```
```suggestion
Rule labels
-----------
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+* :term:`Generic rules <Generic rule>` and :term:`URL callbacks <URL callback>`
+  use their :term:`callable <Plugin callable>`'s name by default (i.e. the
+  function's ``__name__``). This can be overriden with the
```
```suggestion
  function's ``__name__``). This can be overridden with the
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+  function's ``__name__``). This can be overriden with the
+  :func:`sopel.plugin.label` decorator.
+* A :term:`Named rule` is already named (by definition), so it uses its name
+  directly as rule label. This can't be overriden by a decorator.
```
```suggestion
  directly as rule label. This can't be overridden by a decorator.
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+This label is particulary useful for bot owners who want to disable a rule in
+a specific channel. In the following example, the ``say_hello`` rule from the
+``hello`` plugin is disable in the ``#rude`` channel:
```
```suggestion
``hello`` plugin is disabled in the ``#rude`` channel:
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+   [#rude]
+   disable_commands = {'hello': ['say_hello']}
+
+The rule in question is defined by the ``hello`` plugin like that::
```
```suggestion
The rule in question is defined by the ``hello`` plugin like so::
```

Colloquial usage > perfect grammar

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+.. py:function:: plugin_callable(bot, trigger)
+
+   :param bot: the bot's instance
```
```suggestion
   :param bot: wrapped bot instance
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+message that matches the rule.
+
+The ``bot`` provides the ability to send messages to the network (to say
+something or to send a specific command such as ``join``), and to check the
```
```suggestion
something or to send a specific command such as ``JOIN``), and to check the
```

IRC commands are UPPERCASE.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+The ``trigger`` provides information about the line which triggered the rule
+and this callable to be executed.
+
+The return value of a callable is ignored  unless it is
```
```suggestion
The return value of a callable is ignored unless it is
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+The return value of a callable is ignored  unless it is
+:const:`sopel.plugin.NOLIMIT`, in which case
+:term:`rate limiting <Rate limiting>` will not be applied for that call.
+(see :func:`sopel.plugin.rate`)
```
```suggestion
(See :func:`sopel.plugin.rate`.)
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+.. note::
+
+   Note that the name can, and should, be anything, and it doesn't have to be
+   called ``plugin_callable``. At least, it should not be called ``callable``
+   as it is a :func:`Python built-in function <callable>`::
```
```suggestion
   called ``plugin_callable``. At least, it should not be called ``callable``,
   since that is a :func:`Python built-in function <callable>`::
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+===========
+
+Another feature available to plugins is the ability to define
+:term:`Plugin job`. It is a Python callable decorated with
```
```suggestion
:term:`Plugin job`\s. It is a Python callable decorated with
```

Does this work? I didn't test. 🙊

Exirel replied:
I found a way.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+.. py:function:: plugin_job(bot)
+
+   :param bot: the bot's instance
```
```suggestion
   :param bot: the bot instance
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+.. py:function:: setup(bot)
+
+   :param bot: the bot's instance
```
```suggestion
   :param bot: the bot instance
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+appropriate configuration variables exist and are set). Note that this normally
+occurs prior to connection to the server, so the behavior of the messaging
+functions on the :class:`sopel.bot.Sopel` object it's passed is undefined and
+is likely to fail.
```
```suggestion
they are likely to fail.
```

I think this is a copy-paste of old text, but let's fix it anyway.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+This is useful when requiring the presence of configuration values (by raising
+a :exc:`~sopel.config.ConfigurationError` error) or making other environmental
+requirements (dependencies, folder, access right, and so on).
```
```suggestion
requirements (dependencies, file/folder access rights, and so on).
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+.. py:function:: shutdown(bot)
+
+   :param bot: the bot's instance
```
```suggestion
   :param bot: the bot instance
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+   :type bot: :class:`sopel.bot.Sopel`
+
+This function is optional. If it exists, it will be called while the bot
+is shuting down. Note that this normally occurs after closing connection
```
```suggestion
is shutting down. Note that this normally occurs after closing connection
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+This function is optional. If it exists, it will be called while the bot
+is shuting down. Note that this normally occurs after closing connection
+to the server, so the behavior of the messaging functions on the
+:class:`sopel.bot.Sopel` object it's passed is undefined and is likely to fail.
```
```suggestion
:class:`sopel.bot.Sopel` object it's passed is undefined and they are likely to fail.
```

I sense this might cause an orphan line, but oh well.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+:class:`sopel.bot.Sopel` object it's passed is undefined and is likely to fail.
+
+The purpose of this function is to perform whatever actions are needed to allow
+a plugin to properly clean up (e.g, ensuring that any temporary cache files are
```
```suggestion
a plugin to properly clean up (e.g. ensuring that any temporary cache files are
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+A plugin can define and use a configuration section. By subclassing
+:class:`sopel.config.types.StaticSection`, it can define the options it uses
+and may requires. Then, it should add this section to the bot's settings::
```
```suggestion
and may require. Then, it should add this section to the bot's settings::
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+   def setup(bot):
+      bot.settings.define_section('foo', FooSection)
+
+This will allow the bot to properly loads this part of the configuration file:
```
```suggestion
This will allow the bot to properly load this part of the configuration file:
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+Configuration wizard
+--------------------
+
+When the owner setups the bot, Sopel provides a configuration wizard. When a
```
```suggestion
When the owner sets up the bot, Sopel provides a configuration wizard. When a
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+   :param settings: the bot's configuration object
+   :type settings: :class:`sopel.config.Config`
+
+It's intended purpose is to use the methods of the passed
```
```suggestion
Its intended purpose is to use the methods of the passed
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,345 @@
+
+.. note::
+
+   The ``configure`` function is called only from the command line, and a
```
```suggestion
   The ``configure`` function is called only from the command line, and
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -252,16 +252,16 @@ The ``shutdown`` function must follow this interface:
+is shutting down. Note that this normally occurs after closing connection
 to the server, so the behavior of the messaging functions on the
-:class:`sopel.bot.Sopel` object it's passed is undefined and is likely to fail.
+:class:`sopel.bot.Sopel` object it's passed is undefined and they are likely to fail.
```
Well, my suggestion here didn't cause an orphan line, but that's because you just applied the suggestion and didn't re-wrap. 😁

Exirel replied:
I'll, eventually. 😏

dgw commented:
Whipped through here again right before my next meeting. Hopefully I answered all the little sub-threads that needed it, so you can proceed with your next steps. 😸

Exirel commented:
Added some documentation about the ``bot`` parameter for plugin callables. I think it'll be a great place to add information asked by users. A lot of the times we had questions, I wished we could have some place to put the answers, and I think this new structure for the documentation will be the right place.

Exirel commented:
@dgw brace yourself, it's the real one now. I haven't added that many things, and yet, I'm sure you'll find plenty to fix.

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,347 @@
+likely to fail.
+
+The purpose of this function is to perform whatever actions are needed to allow
+a plugin to properly clean up (e.g. ensuring that any temporary cache files are
+deleted).
```
```suggestion
a plugin to properly clean up after itself (e.g. ensuring that any temporary
cache files are deleted).
```

dgw commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,347 @@
+
+Its intended purpose is to use the methods of the passed
+:class:`sopel.config.Config` object in order to create the configuration
+variables it needs to function properly.
```
```suggestion
variables it needs to work properly.
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,135 @@
+=================
+
+Before writing your first plugin, you may want to know what is a plugin, how is
+it made, and what are the best solutions for your need. In this chapter we'll
```
```suggestion
it made, and what are the best solutions for your needs. In this chapter we'll
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,135 @@
+* you don't have any dependencies other than Sopel and the Python's built-in
+  library
+
+This type of plugin is easy to install and share: copy its file into the right
+location, and Sopel will load it on startup.
```
```suggestion
This type of plugin is easy to install: copy its file into the right location,
and Sopel will load it on startup.
```

Goes with my "alternative suggestion" for the above list of reasons to use single-file plugins.

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,135 @@
+Entry point
+-----------
+
+.. versionadded:: 7.0.0
```
```suggestion
.. versionadded:: 7.0
```

I can't think of any places we include a `.z` version in these annotations; it's (pretty much?) always `x.y` only.

Exirel replied:
Oh yeah, why did I do that?

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,135 @@
+* you want a modern and reliable way to package your Sopel plugin
+* you want to distribute more than one Sopel plugin per distributed package
+
+An entry point plugin is the best way to package and distribute a Sopel plugin
+or collection of plugins in a clean and easy-to-update manner.
```
```suggestion
An entry point plugin is the best, most flexible way to package and distribute
a Sopel plugin (or collection of plugins) in a clean, easy-to-update manner.
```

Somehow I knew last time that I'd end up revisiting this one. 😹

Exirel commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,347 @@
+Sopel identifies a callable as a rule when it has been decorated with any of
+these :mod:`sopel.plugin`'s decorators:
+
+* :term:`Generic rule`:: :func:`~sopel.plugin.rule`,
```
:exclamation:

Exirel commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,347 @@
+* per-channel: how often a rule triggers for a given channel
+* globally: how often a rule triggers accross the whole network
+
+Example::
+
+   from sopel import plugin
+
+   @plugin.rule(r'Ah[!?.]?')
+   @plugin.rate(user=2)
+   def you_said_ah(bot, trigger):
+      bot.reply('Ha AH!')
```
Good point!

Exirel commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,347 @@
+
+A callable must accept two positional arguments: a
+:class:`bot <sopel.bot.SopelWrapper>` object, and a
+:class:`trigger <sopel.trigger.Trigger>` object. Both are objects tied to the
+message that matches the rule.
```
"exact" doesn't speak to me.

Exirel commented on docs/source/plugin/bot.rst:
```diff
@@ -0,0 +1,139 @@
+Make it act
+===========
+
+Beside talking, the bot can also **act**:
```
Whaaaaaat?

Exirel commented on docs/source/plugin/bot.rst:
```diff
@@ -0,0 +1,139 @@
+channel like this::
+
+    user_privileges = channel.privileges[trigger.nick]
+    user = channels.users[trigger.nick]
```
`bot.users` is deprecated IIRC.

dgw replied:
Not according to the live docs. https://sopel.chat/docs/bot.html#sopel.bot.Sopel.users

Also not according to the current code: https://github.com/sopel-irc/sopel/blob/528e1b8298e455c486b2a0284bb62a659034f98d/sopel/bot.py#L102-L108

Exirel replied:
OK I'm wrong.

My intent here was to keep that about channel management: both the user and their privileges are from the channel.

dgw replied:
Ah, yeah. Ignore my `bot.users` idea then. The channel property is better here!

Exirel commented on docs/source/plugin/anatomy.rst:
```diff
@@ -0,0 +1,347 @@
+By default anyone can trigger a :term:`rule <Rule>`, and for some it might be
+better to limit who can trigger them. There are decorators for that:
+
+* :func:`sopel.plugin.require_account`: requires services/NickServ
+  authentication; works only if the server implements modern IRC authentication
```
You're totally right!

Exirel commented:
> Did not expect this many line notes in another go-around, but here we are. Have fun! sweat_smile

I added a whole new section, that checks out.

> The singular matches the Sopel submodule that plugin devs work with the most, I guess.

Both work for me, so your call! It's one `git mv` away so pretty easy.

dgw commented:
@Exirel I've seen the fixes since my last review, finally. Go ahead and squash/rebase, then I'll re-review one more time before hopefully merging. There can't be many typos left by now! 😅

Exirel commented:
<loud happy noise>

Exirel commented:
Done!

Exirel commented:
@dgw I've added a new commit with some minor changes, to address the dependencies question. I added that it is better not to have dependencies, because it is more difficult to share the plugin. I added a note to tell plugin authors to write the list of dependencies in a `requirements.txt` file and to document them. I hope this will clarify what I meant.

I still think Folder plugins are not a good idea, as you need to add a directory to the `sys.path`, which is **bad**.
I added the information about dependencies, with the note, so I believe this problem is addressed too.

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,154 @@
+
+A **Folder** plugin is a folder that contains a ``__init__.py`` file. It is
+available to Sopel the same way a :ref:`Single file` plugin is, and shares the
+same limitation:
```
```suggestion
same warnings/advice:
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -52,6 +65,12 @@ When a plugin author wants to split code into different files, and/or share it
 
+.. note::
+
+   As per the single file plugin, you can still use a folder plugin  with extra
+   dependencies, and you can follow the same advices about ``requirements.txt``
+   and ``pip``.
```
```suggestion
   Just like single file plugins, you *can* use extra dependencies with folder
   plugins; the same advice about ``requirements.txt`` and ``pip`` applies.
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -27,22 +27,35 @@ A **Single file** plugin is the most basic form a Sopel plugin can take. It is
 * everything is in a single file
-* you don't have any dependencies other than Sopel and Python's built-in library,
-  or your plugin is a "one-off" for your use only (not intended to be shared)
+* if possible, dependencies should be limited to Sopel and Python's built-in
+  library, or your plugin is a "one-off" for your use only (not intended to be
+  shared)
```
Idea: nest the thing

```suggestion
* either:
  * dependencies are limited to Sopel and Python's built-in library; or
  * your plugin is a "one-off" for your use only (not intended to be shared)
```

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -27,22 +27,35 @@ A **Single file** plugin is the most basic form a Sopel plugin can take. It is
+
+* everything that is not available in this ``__init__.py`` file won't be
+  visible to Sopel
+* if possible, dependencies should be limited to Sopel and Python's built-in
+  library, or your plugin is a "one-off" for your use only (not intended to be
+  shared)
```
```suggestion
* either:
  * dependencies should be limited to Sopel and Python's built-in library,
    for ease of installation; or
  * the plugin is a "one-off", not intended to be shared
```

I'm trying really hard to make this better! 😅

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -27,22 +27,35 @@ A **Single file** plugin is the most basic form a Sopel plugin can take. It is
 
+.. note::
+
+   You can still use a single file plugin with extra dependencies: bot owners
```
```suggestion
   You *can* still use extra dependencies with a single file plugin. Bot owners
```

dgw reviewed (commented):
I'm doing my best to make good suggestions for these new notes now, so you can incorporate them (or get inspired and make your own changes) before going to bed tonight in CEST. 😸 Then I have a prayer of doing that full re-review while you sleep. 😁

Exirel commented:
@dgw if I didn't miss anything, we should be good now!

dgw commented on docs/source/plugin/what.rst:
```diff
@@ -0,0 +1,154 @@
+
+.. seealso::
+
+   The PyPA specification explains in its `Entry points specification`_ what
+   entry points are and how to use them.
```
```suggestion
   The Python Packaging Authority explains how entry points work and how to
   use them in its `Entry points specification`_.
```

I'm mostly avoiding the use of "specification" twice, but expanding the acronym for new Python users who might not know what PyPA is. (Feel free to reject that idea and use the suggestion below instead, which only eliminates the redundant word.)

```suggestion
   The PyPA explains in its `Entry points specification`_ what entry points
   are and how to use them.
```

dgw reviewed (changes requested):
I think we can finish this over the weekend! Home stretch! Glad I did one last full re-review, though. Caught a few stupid typos that I missed the first _n_ times around. x)

Exirel commented on docs/source/plugin/bot.rst:
```diff
@@ -0,0 +1,139 @@
+
+    if user_privileges & plugin.VOICED:
+        # user is voiced
+    elif user_privileges >= plugin.VOICED:
```
Bad typo is bad!

And yes, you can do some nice bitwise operations with privileges, as they are all n^2 values.

Exirel commented:
@dgw I'm ready to rebase & squash asap.

dgw reviewed (approved):
Go for it! 🐑

Exirel commented:
After ~150 comments and many reviews, I've rebased & squashed the whole thing. I'm quite proud of this work, and I hope this will open the path for even greater documentation in the future.

Ready to merge!
